### PR TITLE
plugin/forward: erase expired connections by timer

### PR DESCRIPTION
**Pull request by rdrozhdzh**
### 1. What does this pull request do?
 - in previous implementation, the expired connections resided in
   cache until new request to the same upstream/protocol came. In
   case if the upstream was unhealthy new request may come long time
   later or may not come at all. All this time expired connections
   held system resources (file descriptors, ephemeral ports). In my
   fix the expired connections and related resources are released
   by timer
 - decreased the complexity of taking connection from cache. The list
   of connections is treated as stack (LIFO queue), i.e. the connection
   is taken from the end of queue (the most fresh connection) and
   returned to the end (as it was implemented before). The remarkable
   thing is that all connections in the stack appear to be ordered by
   'used' field
 - the cleanup() method finds the first good (not expired) connection
   in stack with binary search, since all connections are ordered by
   'used' field

### 2. Which issues (if any) are related?
#1768

### CODE REVIEW

File: plugin/forward/persistent.go
Comment (miekg) [inline plugin/forward/persistent.go:110]: Doing this in a goroutine is a data race on conns, no?

File: plugin/forward/persistent.go
Comment (miekg) [inline plugin/forward/persistent.go:134]: So you're moving the complexity to this function, which would be O(n*log(n)) I guess, best case. I opted not doing this because we just want a conn that is valid, not the last one used...

**Review by miekg — CHANGES_REQUESTED** [review]
Thanks for picking this up. This bugged me as well, as my local monitoring showed 4 conns that never expired.

I think we don't really need the Sort. The closing is more tricky, as you don't want to hold up giving out new connections...

File: plugin/forward/persistent.go
Comment (rdrozhdzh) [reply plugin/forward/persistent.go:110]: I remove these connections from `transport.conns` before calling `closeConns()`, so they are unreachable from transport anymore

File: plugin/forward/persistent.go
Comment (rdrozhdzh) [reply plugin/forward/persistent.go:134]: > So you're moving the complexity to this function, which would be O(n*log(n))

I believe the complexity of Search() is O(log(n)). So, this should be faster than iterating over all cached connections.

> we just want a conn that is valid, not the last one used...

yes, and the fastest way to find a valid connection is to check the last one. Since it was added last, it has higher probability (comparing to other connections) that it is not expired. If the last (in the slice) connection is expired this means all previous connections are expired as well, we even don't need to check them

Another benefit of taking connection from the end of slice is that we would potentially process higher traffic with less connections in cache (i.e. consume less system resources). Lets consider the following case.

Expiration time is 10s. We have 100 connections in cache after some previous user activity. At some point user starts sending 1 DNS query each 100ms. The DNS processing time is 10ms.

The original code will take a cached connection **from the begin** of cached conns and 10ms later it will add the connection back to the end of conns list with updated "used" timestamp. So, it will just rotate the 100 connections in the cache without releasing them, because each connection will be periodically updated with new timestamp, and no connection will expire

The new code will take a connection **from the end** of list and 10ms later the connection will be added back to the end of conns list with updated "used" timestamp. On next request we will use same connection again. After 10s all connections except the last one will get expired and released (by timer). Only one connection will reside in cache and this single connection would be enough to process current load (i.e. 1 query per 100ms)

So, I would say the new approach is more adaptive to the load and more prudent about system resources

**Comment by rdrozhdzh** [comment]
@miekg, please clarify which changes you want me to add here

**Comment by miekg** [comment]
sorry to be late; busy w/ work lately. Looking now.

your replies are convincing

File: plugin/forward/persistent.go
Comment (miekg) [reply plugin/forward/persistent.go:110]: But `conns` is and because we're only `ranging` over it; that makes it save?

File: plugin/forward/persistent.go
Comment (miekg) [reply plugin/forward/persistent.go:134]: ok, trust you on this.

File: plugin/forward/persistent.go
Comment (rdrozhdzh) [reply plugin/forward/persistent.go:110]: `conns` here is just a parameter passed to `closeConns()`, i.e. local variable. This is not the same as `transport.conns`

Comment (miekg) [reply plugin/forward/persistent.go:110]: it's a slice, so the underlaying array is shared between the two; it still don't think this is safe to do...

Comment (miekg) [reply plugin/forward/persistent.go:110]: To move this forward, can we not expire and sort in a separate goroutine? (We do need to quantify the potential stall then though)

Comment (rdrozhdzh) [reply plugin/forward/persistent.go:110]: yes, the `conns` is a slice, but the underlaying array is not shared. The second reference to this array is erased from `transport.conns` right before calling `closeConns()`, see the lines
https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR71
and
https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR133

And seems I missed the line
https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR122
which also should be moved before `closeConns()`

We actually don't expire and search connections in a separate goroutine. The `cleanup()` is called synchronously from `connManager()`. We only call  `closeConns()` in a separate goroutine which just closes connections which are not accessible from `connManager()` anymore

Comment (rdrozhdzh) [reply plugin/forward/persistent.go:110]: Well, strictly speaking, in case of https://github.com/coredns/coredns/pull/1782/files#diff-c641d6db9fa84aad4501394efccea2dbR133 the same underlying array is still used, but slices are not intersected. I.e. `connManager()` has a slice with no references to the connections passed to `closeConns()` with another slice. Adding new connection to cache cannot affect the connections passed to `closeConns()`

Comment (miekg) [reply plugin/forward/persistent.go:110]: ok

**Comment by miekg** [comment]
@johnbelamaric do you have some spare cycles to take a look here. 

The final conclusion looks to be this is OK, but I like a second opinion.

File: plugin/forward/persistent.go
Comment (johnbelamaric) [reply plugin/forward/persistent.go:110]: I don't think that's true. I *think* the little program below will simulate this situation:

```
package main

import (
        "log"
        "time"
)

func close(s []int) {
        log.Println("Pre-sleep closing: ", s)
        time.Sleep(1*time.Second)
        log.Println("Post-sleep closing: ", s)
}

func add(s []int) {
}

func main() {
        log.Println("Hi")

        s := make([]int,10)
        for i := 0; i < 10; i++ {
                s[i] = i
        }

        ss := s[5:]
        s = s[:5]
        go close(ss)

        time.Sleep(time.Second/2)
        s = append(s,100)

        time.Sleep(2*time.Second)
        log.Println("s = ", s)
}
```

Running this we see:

```
jbelamaric@ANN-M-JBELAMARIC [k8s:minikube] ~/slices$ go run main.go
2018/05/23 18:28:30 Hi
2018/05/23 18:28:30 Pre-sleep closing:  [5 6 7 8 9]
2018/05/23 18:28:31 Post-sleep closing:  [100 6 7 8 9]
2018/05/23 18:28:32 s =  [0 1 2 3 4 100]
jbelamaric@ANN-M-JBELAMARIC [k8s:minikube] ~/slices$
```

Is there something different about that and the situation here?

Comment (rdrozhdzh) [reply plugin/forward/persistent.go:110]: @johnbelamaric , your case is different

I pass to the `closeConns()` goroutine the slice which points to the beginning part of the array. And the slice pointing to remaining part overwrites the original slice in transport object which previously pointed to whole array . So, the transport cannot access the elements which resided in the beginning of the original array anymore. If transport appends new element it will be either added to the end of original array, or added to a newly allocated array after copying elements which were accessible from the transport slice. In any case the elements passed to `closeConns()` will remain in place and will not change.

To simulate this situaltion, you need to replace
```
        ss := s[5:]
        s = s[:5]
```
in your example with 
```
        ss := s[:5]
        s = s[5:]
```

Comment (johnbelamaric) [reply plugin/forward/persistent.go:110]: Ah, got it. Ok 👍

Comment (miekg) [reply plugin/forward/persistent.go:110]: Thanks both.

We're good to go then?

Comment (miekg) [reply plugin/forward/persistent.go:110]: I would like some comments in the code on this; it's not *that* obvious how the race is avoided.

Comment (rdrozhdzh) [reply plugin/forward/persistent.go:110]: done

**Comment by miekg** [comment]
@johnbelamaric happy as well?

**Review by miekg — APPROVED** [review]

**Review by johnbelamaric — APPROVED** [review]
👍

**Comment by johnbelamaric** [comment]
Pending seeing the CI work!

**Pull request merged**
